lib/Reader/readerir.cpp

```diff
@@ -366,17 +379,47 @@ void GenIR::readerPrePass(uint8_t *Buffer, uint32_t NumBytes) {
       continue;
     }
 
-    const ABIArgInfo ArgInfo = ABIMethodSig.getArgumentInfo(J);
-    if (ArgInfo.getKind() == ABIArgInfo::Indirect) {
+    const ABIArgInfo &ArgInfo = ABIMethodSig.getArgumentInfo(J);
+    switch (ArgInfo.getKind()) {
+    case ABIArgInfo::Indirect:
       // Indirect arguments aren't homed.
-    } else {
-      Value *Arg = &*ArgI;
-      Value *Home = Arguments[J];
-      Type *HomeType = Home->getType()->getPointerElementType();
-      Arg = ABISignature::coerce(*this, HomeType, Arg);
-      const bool IsVolatile = false;
-      storeAtAddressNoBarrierNonNull((IRNode *)Home, (IRNode *)Arg, HomeType,
-                                     IsVolatile);
+      assert(!doesArgumentHaveHome(ArgInfo));
+      break;
+
+    case ABIArgInfo::Expand: {
+      Type *BytePtrType = Type::getInt8PtrTy(LLVMContext, 0);
+      Value *Home = LLVMBuilder->CreatePointerCast(Arguments[J], BytePtrType);
+
+      ArrayRef<ABIArgInfo::Expansion> Expansions = ArgInfo.getExpansions();
+      uint32_t ExpansionCount = Expansions.size();
+      for (uint32_t E = 0; E < ExpansionCount - 1; ++E, ++I, ++ArgI) {
+        ABISignature::collapse(*this, Expansions[E], &*ArgI, Home);
+      }
+      ABISignature::collapse(*this, Expansions[ExpansionCount - 1], &*ArgI,
+                             Home);
```

Comment:
The containing loop (i.e. the loop over the argument infos) will also increment `I` and `ArgI`. This last `collapse` is outside the inner loop to account for that increment.